### PR TITLE
Jupyterhub chart

### LINKED ISSUE #152: include jupyterhub in helm charts
add and configure jupyterhub in the helm charts (see e.g. https://github.com/jupyterhub/binderhub/tree/master/helm-chart) using our spawner. This is a continuation of #147 

---

addresses #146

### PR DESCRIPTION
**Pull request by rokroskar**
This PR adds jupyterhub to the k8s chart.

## To-do: 

- [x] add jupyterhub client to the gitlab client db
- [x] remove dynamic PVCs
- [x] use postgres and not sqlite
- [x] include the sidekick /projects service
- [x] use the renga spawner mixin
- [x] tested notebook spawn on k8s
- [x] tested notebook spawn with docker-compose (with jupyterhub 0.8.1, see #194)

## To use

Follow the directions in`charts/README.rst`

## Breaking changes

The notebooks are now launched on `/services/projects/<namespace>/<project>/<commit-sha>`. This change **breaks the current functionality of the UI** (cc @ableuler, @ciyer) which looks at the project's environments as defined in gitlab to provide the appropriate link for launching the notebook. Hopefully this is a minor fix. 

## Outstanding issues

1. A bug in the pre-puller causes helm to occasionally hang. The prepuller job needs to be deleted and helm moves on -- however, I disabled the prepuller altogether in the chart. This has been fixed in JH charts (see https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/477). To use it, we will need to update the JH chart version we use -- this will most easily be done when we refactor the JH chart we have here to use it as a simple dependency and move the initialization steps to a helm post-install script. 
The prepuller bug might only be a problem on minikube, in which case turning it off should be moved to `minikube-values.yaml`. 

2. The environment for the notebook container is grabbed from the `get_env` method of the spawner and cannot easily be amended otherwise. This means that at the moment we cannot put the repository URL into the environment to push the notebook back to the repo. The fix will need to address the outstanding issue of getting the repository (async) method result inside the `get_env` method. I will open an issue and perhaps @jirikuncar can handle this? 

---

addresses #146 
closes #152

### CODE REVIEW

**Comment by leafty** [comment]
Now, access the hub with `http://$(minikube ip)/jupyterhub/` (trailing slash needed)

**Comment by leafty** [comment]
to do:
- [ ] remove dynamic PVCs
- [ ] use postgres and not sqlite
- [ ] use the renga spawner mixin
- [ ] include the sidekick /projects service

**Comment by leafty** [comment]
Note: `postgres` won't init `jupyterhub` database if it exists already.
Either `helm delete --purge` or manually run:
```bash
$ kubectl -n renga exec -it renga-staging-renga-postgresql-[...] psql
root@renga-staging-renga-postgresql-59f5fd666f-ch69j:/# cd docker-entrypoint-initdb.d/
root@renga-staging-renga-postgresql-59f5fd666f-ch69j:/docker-entrypoint-initdb.d# . init_jupyterhub_db.sh
```

**Comment by rokroskar** [comment]
to use the custom image, build using the `jupyterhub-k8s.Dockerfile` in `services/jupyterhub` and tag with `rengahub/jupyterhub-k8s:latest` or whatever you specify in the `values.yaml`

**Comment by jirikuncar** [comment]
- [ ] enable Auto-DevOps for GitLab
- [ ] use sub-domains for services

**Comment by rokroskar** [comment]
@jirikuncar I want to keep those for another PR -- keep this one focused on JupyterHub. Can you move the comment above into an issue?

**Comment by jirikuncar** [comment]
Sure, I can move it, but please note that without the Auto-DevOps enabled one has to manually create environments on GitLab.

File: charts/renga/charts/gitlab/values.yaml [hunk 1]
```diff
@@ -3,8 +3,8 @@
 # Declare variables to be passed into your templates.
 
 image:
-  repository: gitlab/gitlab-ce
-  tag: latest
+  repository: rengahub/gitlab
+  tag: development
```
Comment (leafty) [inline charts/renga/charts/gitlab/values.yaml]: can't we go with the standard image?

File: docker-compose.yml [hunk 2]
```diff
@@ -83,7 +83,8 @@ services:
       retries: 5
 
   gitlab:
-    image: gitlab/gitlab-ce:latest
+    image: ${DOCKER_PREFIX}gitlab:${PLATFORM_VERSION}
+    build: services/gitlab
```
Comment (leafty) [inline docker-compose.yml]: out-of-scope changes (dockerfile), put then in another PR

File: charts/renga/charts/gitlab/values.yaml (hunk 1 above)
Comment (jirikuncar) [reply charts/renga/charts/gitlab/values.yaml]: no we can't for now

File: docker-compose.yml (hunk 2 above)
Comment (jirikuncar) [reply docker-compose.yml]: It's related so the environments are created properly for jupyterhub access control checks.

File: charts/renga/charts/gitlab/values.yaml (hunk 1 above)
Comment (leafty) [reply charts/renga/charts/gitlab/values.yaml]: yes you can, you just need to paste the content of `Auto-DevOps.gitlab-ci.yml` in a cm and mount it...

File: docker-compose.yml (hunk 2 above)
Comment (leafty) [reply docker-compose.yml]: the `docker-compose.yml` has ZERO effect on the charts

File: charts/README.rst [hunk 3]
```diff
@@ -10,9 +10,33 @@ Requires minikube, kubectl and helm.
 
     $ minikube start
     $ helm init
+    # Run only once!
+    $ echo "$(minikube ip) renga-k8s.build keycloak.renga-k8s.build gitlab.renga-k8s.build jupyterhub.renga-k8s.build" | sudo tee -a /etc/hosts
```
Comment (leafty) [inline charts/README.rst]: scope this when #172 is solved

File: charts/renga/values.yaml [hunk 4]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (leafty) [inline charts/renga/values.yaml]: Do not need the spaces

File: charts/README.rst (hunk 3 above)
Comment (rokroskar) [reply charts/README.rst]: it's not doing much apart from setting a host for the IP -- AutoDevOps doesn't work without it.

**Comment by leafty** [comment]
```bash
[C 2018-05-03 06:57:30.307 JupyterHub application:90] Bad config encountered during initialization:
[C 2018-05-03 06:57:30.307 JupyterHub application:91] The 'spawner_class' trait of <jupyterhub.app.JupyterHub object at 0x7f0b481dc550> instance must be a type, but 'spawners.RengaKubeSpawner' could not be importe
```

**Comment by rokroskar** [comment]
@leafty should be fixed in d26f192

**Comment by leafty** [comment]
```bash
$ helm install --name renga --namespace renga -f minikube-values.yaml renga --set jupyterhub.rbac.enabled=true --wait --timeout 600
Error: release renga failed: timed out waiting for the condition
```

**Comment by leafty** [comment]
```
==> /var/log/gitlab/unicorn/unicorn_stdout.log <==
I, [2018-05-03T08:03:06.166246 #598]  INFO -- omniauth: (oauth2_generic) Callback phase initiated.

==> /var/log/gitlab/gitlab-rails/production.log <==

Faraday::ConnectionFailed (Failed to open TCP connection to renga-k8s.build:80 (getaddrinfo: Name or service not known)):
  lib/gitlab/middleware/multipart.rb:95:in `call'
  lib/gitlab/request_profiler/middleware.rb:14:in `call'
  lib/gitlab/middleware/go.rb:17:in `call'
  lib/gitlab/etag_caching/middleware.rb:11:in `call'
  lib/gitlab/middleware/read_only/controller.rb:28:in `call'
  lib/gitlab/middleware/read_only.rb:16:in `call'
  lib/gitlab/request_context.rb:18:in `call'
  lib/gitlab/metrics/requests_rack_middleware.rb:27:in `call'
  lib/gitlab/middleware/release_env.rb:10:in `call'



==> /var/log/gitlab/gitlab-workhorse/current <==
2018-05-03_08:03:06.17650 renga-k8s.build @ - - [2018/05/03:08:03:06 +0000] "GET /gitlab/users/auth/oauth2_generic/callback?state=f60917694ca3132a9d09c40119e53bc04bff8bd24b8280a4&session_state=5c95678c-6af0-4c98-9107-16a779b7c9dc&code=eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0..k29dQvz5YJcjMJwh7aBQKw.FNz8PTYj97813qUCq5al-gkXZMrpQlgGVn8LtDZmsDdwoERR9x_vqco5g8lvR5ut1q7-i1jvyEXfMAoG7u2OkqzyUPP_z5Kqows2vxX0gLk5gF5-MExE-Jgbb-1vQIgXp5zQ3UXKBqdXU-mCGhwOrSfyVEkWpSsJKC5NZa8-dadJrWhuj_kVWUl35PLJmmI5uHBwmssLyXWHWVFH78aJPYMMnp8dVSd4GndXIObPI2W-tdzO8jArMZLtmVbYHnsM.iAGEUKpJjQtzA25WQte9SQ HTTP/1.1" 500 2902 "http://renga-k8s.build/gitlab/users/sign_in" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36 OPR/52.0.2871.99" 0.014

==> /var/log/gitlab/nginx/gitlab_access.log <==
172.17.0.1 - - [03/May/2018:08:03:06 +0000] "GET /gitlab/users/auth/oauth2_generic/callback?state=f60917694ca3132a9d09c40119e53bc04bff8bd24b8280a4&session_state=5c95678c-6af0-4c98-9107-16a779b7c9dc&code=eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0..k29dQvz5YJcjMJwh7aBQKw.FNz8PTYj97813qUCq5al-gkXZMrpQlgGVn8LtDZmsDdwoERR9x_vqco5g8lvR5ut1q7-i1jvyEXfMAoG7u2OkqzyUPP_z5Kqows2vxX0gLk5gF5-MExE-Jgbb-1vQIgXp5zQ3UXKBqdXU-mCGhwOrSfyVEkWpSsJKC5NZa8-dadJrWhuj_kVWUl35PLJmmI5uHBwmssLyXWHWVFH78aJPYMMnp8dVSd4GndXIObPI2W-tdzO8jArMZLtmVbYHnsM.iAGEUKpJjQtzA25WQte9SQ HTTP/1.1" 500 2902 "http://renga-k8s.build/gitlab/users/sign_in" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36 OPR/52.0.2871.99"
```

**Comment by rokroskar** [comment]
@jirikuncar @leafty - this works now with the caveats in the PR description.

File: services/jupyterhub/spawners.py [hunk 5]
```diff
@@ -71,24 +70,24 @@ def get_env(self):
...
         self.log.info(
             "starting with args: {}".format(' '.join(self.get_args()))
         )
-        self.log.info("user options: {}".format(self.user_options))
+        self.log.debug("user options: {}".format(self.user_options))
 
         auth_state = yield self.user.get_auth_state()
         assert 'access_token' in auth_state
-        self.log.info(auth_state)
+        self.log.debug(auth_state)
```
Comment (jirikuncar) [inline services/jupyterhub/spawners.py]: we should remove it for security reasons

File: services/jupyterhub/spawners.py [hunk 6]
```diff
@@ -187,11 +179,12 @@ def start(self):
                 entrypoint='sh -c',
                 command=[
                     'git clone {repository} {volume_path} && '
-                    'git checkout -b {environment_slug} {commit_sha} && '
+                    'git checkout -b {branch}-{username} {commit_sha} && '
```
Comment (jirikuncar) [inline services/jupyterhub/spawners.py]: I would use only `{branch}` in this case.

File: services/jupyterhub/spawners.py [hunk 7]
```diff
@@ -253,42 +246,56 @@ def get_pod_manifest(self):
...
                 'args': [
-                    'clone',
-                    '--single-branch',
-                    '-b',
-                    self.git_revision,
-                    '--',
-                    repository,
-                    '/repo',
+                    'git clone {repository} {mount_path} && '
+                    'git checkout -b {branch}-{username} {commit_sha} '.format(
```
Comment (jirikuncar) [inline services/jupyterhub/spawners.py]: only `{branch}`?

File: services/jupyterhub/spawners.py (hunk 6 above)
Comment (rokroskar) [reply services/jupyterhub/spawners.py]: then you can't check out a specific commit (branch already exists). We could just checkout the specific commit and not be on a branch at all I suppose.
